okhttp/src/main/java/okhttp3/internal/http/RetryAndFollowUpInterceptor.kt

```diff
@@ -0,0 +1,330 @@
+    // proxy and would manifest as a standard IOException. Unless it is one we know we should not
+    // retry, we return true and try a new route.
+    return true
+  }
+
+  /**
+   * Figures out the HTTP request to make in response to receiving `userResponse`. This will
+   * either add authentication headers, follow redirects or handle a client request timeout. If a
+   * follow-up is either unnecessary or not applicable, this returns null.
+   */
+  @Throws(IOException::class)
+  private fun followUpRequest(userResponse: Response, route: Route?): Request? {
+    val responseCode = userResponse.code()
+
+    val method = userResponse.request().method()
+    when (responseCode) {
+      HTTP_PROXY_AUTH -> {
+        val selectedProxy = route?.proxy() ?: client.proxy()
+        if (selectedProxy?.type() != Proxy.Type.HTTP) {
+          throw ProtocolException("Received HTTP_PROXY_AUTH (407) code while not using proxy")
+        }
+        return client.proxyAuthenticator().authenticate(route, userResponse)
+      }
+
+      HTTP_UNAUTHORIZED -> return client.authenticator().authenticate(route, userResponse)
+
+      HTTP_PERM_REDIRECT, HTTP_TEMP_REDIRECT -> {
+        // "If the 307 or 308 status code is received in response to a request other than GET
+        // or HEAD, the user agent MUST NOT automatically redirect the request"
+        if (method != "GET" && method != "HEAD") {
+          return null
+        }
+        return buildRedirectRequest(userResponse, method)
+      }
+
+      HTTP_MULT_CHOICE, HTTP_MOVED_PERM, HTTP_MOVED_TEMP, HTTP_SEE_OTHER -> {
+        return buildRedirectRequest(userResponse, method)
+      }
+
+      HTTP_CLIENT_TIMEOUT -> {
+        // 408's are rare in practice, but some servers like HAProxy use this response code. The
+        // spec says that we may repeat the request without modifications. Modern browsers also
+        // repeat the request (even non-idempotent ones.)
+        if (!client.retryOnConnectionFailure()) {
+          // The application layer has directed us not to retry the request.
+          return null
+        }
+
+        val requestBody = userResponse.request().body()
+        if (requestBody != null && requestBody.isOneShot()) {
+          return null
+        }
+        val priorResponse = userResponse.priorResponse()
+        if (priorResponse != null && priorResponse.code() == HTTP_CLIENT_TIMEOUT) {
+          // We attempted to retry and got another timeout. Give up.
+          return null
+        }
+
+        if (retryAfter(userResponse, 0) > 0) {
+          return null
+        }
+
+        return userResponse.request()
+      }
+
+      HTTP_UNAVAILABLE -> {
+        val priorResponse = userResponse.priorResponse()
+        if (priorResponse != null && priorResponse.code() == HTTP_UNAVAILABLE) {
+          // We attempted to retry and got another timeout. Give up.
+          return null
+        }
+
+        if (retryAfter(userResponse, Integer.MAX_VALUE) == 0) {
```

Comment:
Should we use Int references here?

Reply:
Once we're all migrated over we'll need a couple passes to use Kotlin things wherever they're preferred.